arviz/kde_utils.py

```diff
@@ -0,0 +1,812 @@
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = 0
+        if bw <= 0:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge to a positive value. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+            return bw
+        if tol >= 1:
+            warnings.warn(
+                "Improved Sheather-Jones did not converge. "
+                "Using Silverman's rule instead.",
+                Warning
+            )
+            bw = (bw_silverman(x) / np.ptp(x)) ** 2
+    return bw
+
+# KDE Utilities ---------------------------------------------------------------------
+
+def check_type(x):
+    """
+    Checks the input is of the correct type.
+    It only accepts numeric lists/numpy arrays of 1 dimension.
+    If input is not of the correct type, an informative message is thrown.
+
+    Parameters
+    ----------
+    x : Object whose type is checked before computing the KDE.
+
+    Returns
+    -------
+    x : 1-D numpy array
+        If no error is thrown, a 1 dimensional array of
+        sample data from the variable for which a density estimate is desired.
+
+    """
+    if not isinstance(x, (list, np.ndarray)):
+        raise ValueError((
+            f"`x` is of the wrong type.\n"
+            f"Can't produce a density estimator for {type(x)}.\n"
+            f"Please input a numeric list or numpy array."
+        ))
+
+    # Will raise an error if `x` can't be casted to numeric
+    x = np.asfarray(x)
+
+    x = x[np.isfinite(x)]
+    if x.size == 0:
+        raise ValueError("`x` does not contain any finite number.")
+
+    if x.ndim != 1:
+        raise ValueError((
+            f"Unsupported dimension number.\n"
+            f"Density estimator only works with 1-dimensional data, not {x.ndim}."
+        ))
+    return x
+
+
+def check_custom_lims(custom_lims, x_min, x_max):
+    """
+    Checks whether `custom_lims` are of the correct type.
+    It accepts numeric lists/tuples of length 2.
+
+    Parameters
+    ----------
+    custom_lims : Object whose type is checked.
+
+    Returns
+    -------
+    None: Object of type None
+
+    """
+    if not isinstance(custom_lims, (list, tuple)):
+        raise TypeError((
+            f"`custom_lims` must be a numeric list or tuple of length 2.\n"
+            f"Not an object of {type(custom_lims)}."
+        ))
+
+    if len(custom_lims) != 2:
+        raise AttributeError(
+            f"`len(custom_lims)` must be 2, not {len(custom_lims)}.")
+
+    any_bool = any(isinstance(i, bool) for i in custom_lims)
+    if any_bool:
+        raise TypeError(
+            "Elements of `custom_lims` must be numeric or None, not bool.")
+
+    if custom_lims[0] is None:
+        custom_lims[0] = x_min
+
+    if custom_lims[1] is None:
+        custom_lims[1] = x_max
+
+    types = (int, float, np.integer, np.float)
```

Comment:
maybe just combine this with the next line.

Reply:
Something like this?

```python
 all_numeric = all(isinstance(i, (int, float, np.integer, np.float)) for i in custom_lims)
```